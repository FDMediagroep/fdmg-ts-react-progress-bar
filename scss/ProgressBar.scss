.progress {
  height: 30px;
  padding: 0 5px;
  .counters {
    display: flex;
    justify-content: space-between;
    padding-top: 10px;
    >span {
      font-size: 10px;
      margin-top: 10px;
      color: rgba(152, 152, 152, 1);
      height: 11px;
      &.hidden {
        display: none;
      }
      &.visible {
        display: inherit;
      }
    }
  }
  .time-line {
    position: relative;
    &:before {
      content: "";
      position: absolute;
      top: 12px;
      border-radius: 5px;
      width: 100%;
      height: 3px;
      display: block;
      background-color: rgba(255,255,255,0.1);
      box-shadow: inset 0 0 1px rgba(0,0,0,0.35);
    }
    .time-line-progress-bar {
      position:absolute;
      top: 12px;
      height: 3px;
      display: block;
      width: 0;
      background-color: #ffd200;
      border-top-left-radius: 5px;
      border-bottom-left-radius: 5px;
      &:after {
        content:"";
        position: absolute;
        right: -4px;
        height: 9px;
        width: 9px;
        border-radius: 50%;
        background-color: #ffd200;
        border: transparent;
        margin-top: 0;
        top: -3px;
      }
    }
    input[type="range"] {
      -webkit-appearance: none;
      -moz-appearance: none;
      appearance: none;
      width: 100%;
      position: absolute;
      background: transparent;
      margin: 0;
      padding: 0;
      opacity: 0;
      height: 30px;
      &:hover {
        cursor: pointer;
      }
      &:focus {
        outline: none;
      }
      &::-webkit-slider-thumb {
        position: relative;
        -webkit-appearance: none;
        border-radius: 50%;
        background-color: #ffd200;
        border: transparent;
        cursor: pointer;
        margin-top: 2px;
        top: 0;
        height: 20px;
        width: 20px;
        opacity: 0;
      }
      &::-moz-range-thumb {
        -moz-appearance: none;
        position: absolute;
        border-radius: 50%;
        background-color: #ffd200;
        border: transparent;
        cursor: pointer;
        height: 20px;
        width: 20px;
        opacity: 0;
      }
      &::-moz-range-track {
        -moz-appearance: none;
        display: none;
        background-color: transparent;
        cursor: pointer;
      }
      &::-ms-thumb, &::-ms-track, &::-ms-fill-upper,  &::-ms-fill-lower  {
        opacity: 0;
      }
    }
  }
}