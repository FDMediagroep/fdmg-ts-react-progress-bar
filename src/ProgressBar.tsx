import * as React from 'react';

export interface Props {
    currentTime: number;
    duration: number;
    percentage: number;
    hideProgressBarCurrentTime: boolean;
    hideProgressBarDuration: boolean;
    onElapsedTimeUpdate: (...args: any[]) => void;
    autoPlay: boolean;
    isBuffering: boolean;
    hideTimeLine: RegExpMatchArray|boolean;
}
/**
 * Progress bar component.
 */
export default class ProgressBar extends React.Component<Props, any> {
    public state: any;
    public props: Props;

    private progressBar: any;

    constructor(props: Props) {
        super(props);
        this.props = props;
    }

    /**
     *  Convert seconds (number) to a human readable time format: 2u 3m 34s
     */
    convertToReadableTime = (seconds: number) => {
        const h = Math.floor(seconds / 3600);
        const m = Math.floor(seconds % 3600 / 60);
        const s = Math.floor(seconds % 3600 % 60);

        const hDisplay = h > 0 ? h + "u " : "";
        const mDisplay = m > 0 ? m + "m " : "";
        const sDisplay = s > 0 ? s + "s" : "";

        return hDisplay + mDisplay + sDisplay;
    }

    render() {
        console.log(this.props.hideProgressBarCurrentTime, this.props.hideProgressBarDuration);
        const spinner = this.props.isBuffering ? (
            <div className="loader">
                <div className="double-bounce1"/>
                <div className="double-bounce2"/>
            </div>
        ) : "";

        const timeLine = this.props.hideTimeLine ? "" : (
            <div className={"time-line"} >
                <span className="time-line-progress-bar"  style={{width : this.props.percentage + "%"}} >
                    {spinner}
                </span>
                <input
                    ref={(progressBar) => { this.progressBar = progressBar; }}
                    type="range"
                    step="0.01"
                    value={this.props.percentage}
                    onInput={this.props.onElapsedTimeUpdate}
                    onChange={this.props.onElapsedTimeUpdate}
                />
            </div>
        );

        const currentTime = this.props.hideProgressBarCurrentTime ? "" : (
            <span>
                {(this.props.currentTime === 0) ? "" : this.convertToReadableTime(this.props.currentTime)}
            </span>
        );

        const duration = this.props.hideProgressBarDuration ? "" : (
            <span>
                {(this.props.duration === 0) ? "" : this.convertToReadableTime(this.props.duration)}
            </span>
        );

        const counters = this.props.hideProgressBarCurrentTime && this.props.hideProgressBarDuration ? "" : (
            <div className="counters">
                {currentTime}
                {duration}
            </div>
        );

        return (
            <div className="progress">
                <div>
                    {timeLine}
                </div>
                {counters}
            </div>
        );
    }
}
