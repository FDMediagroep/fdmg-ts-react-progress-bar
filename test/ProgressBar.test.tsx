import * as React from 'react';
import * as Enzyme from 'enzyme';
import { shallow, mount, render } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

import ProgressBar from "../src/ProgressBar";
import toJson from 'enzyme-to-json';

console.info = () => {};

describe('ProgressBar', () => {
    const empty = () => {};

    beforeAll(() => {
        Enzyme.configure({ adapter: new Adapter() });
    });

    test('renders correctly', () => {
        const component = shallow(
            <ProgressBar
                percentage={33}
                currentTime={1234}
                duration={12345}
                onElapsedTimeUpdate={empty}
                autoPlay={false}
                isBuffering={false}
                hideTimeLine={false}
                hideProgressBarCurrentTime={false}
                hideProgressBarDuration={false}
            />);

        expect(toJson(component)).toMatchSnapshot();
    });

    test('renders correctly without time-line', () => {
        const component = shallow(
            <ProgressBar
                percentage={33}
                currentTime={1234}
                duration={12345}
                onElapsedTimeUpdate={empty}
                autoPlay={false}
                isBuffering={false}
                hideTimeLine={true}
                hideProgressBarCurrentTime={false}
                hideProgressBarDuration={false}
            />);

        expect(toJson(component)).toMatchSnapshot();
    });

    test('renders correctly without the progress bar its currentTime', () => {
        const component = shallow(
            <ProgressBar
                percentage={33}
                currentTime={1234}
                duration={12345}
                onElapsedTimeUpdate={empty}
                autoPlay={false}
                isBuffering={false}
                hideTimeLine={false}
                hideProgressBarCurrentTime={true}
                hideProgressBarDuration={false}
            />);

        expect(toJson(component)).toMatchSnapshot();
    });

    test('renders correctly without the progress bar its duration', () => {
        const component = shallow(
            <ProgressBar
                percentage={33}
                currentTime={1234}
                duration={12345}
                onElapsedTimeUpdate={empty}
                autoPlay={false}
                isBuffering={false}
                hideTimeLine={false}
                hideProgressBarCurrentTime={false}
                hideProgressBarDuration={true}
            />);

        expect(toJson(component)).toMatchSnapshot();
    });

    test('renders correctly when all time indicators are disabled', () => {
        const component = shallow(
            <ProgressBar
                percentage={33}
                currentTime={1234}
                duration={12345}
                onElapsedTimeUpdate={empty}
                autoPlay={false}
                isBuffering={false}
                hideTimeLine={false}
                hideProgressBarCurrentTime={true}
                hideProgressBarDuration={true}
            />);

        expect(toJson(component)).toMatchSnapshot();
    });

    test('input should handle onChange event', () => {
        const updateTrackTime = jest.fn((e) => {
            expect(e.target.value).toBe(75);
        });
        const progressBarComponent = mount(
            <ProgressBar
                percentage={33}
                currentTime={50}
                duration={100}
                onElapsedTimeUpdate={updateTrackTime}
                autoPlay={false}
                isBuffering={false}
                hideTimeLine={false}
                hideProgressBarCurrentTime={false}
                hideProgressBarDuration={false}
            />);

        progressBarComponent.find('input').simulate('change', { target: { value: 75 }});
        expect(updateTrackTime).toHaveBeenCalled();
    });
});
