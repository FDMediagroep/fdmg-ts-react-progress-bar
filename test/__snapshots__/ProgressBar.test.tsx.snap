// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar renders correctly 1`] = `
<div
  className="progress"
>
  <div>
    <div
      className="time-line"
    >
      <span
        className="time-line-progress-bar"
        style={
          Object {
            "width": "33%",
          }
        }
      />
      <input
        onChange={[Function]}
        onInput={[Function]}
        step="0.01"
        type="range"
        value={33}
      />
    </div>
  </div>
  <div
    className="counters"
  >
    <span>
      20m 34s
    </span>
    <span>
      3u 25m 45s
    </span>
  </div>
</div>
`;

exports[`ProgressBar renders correctly when all time indicators are disabled 1`] = `
<div
  className="progress"
>
  <div>
    <div
      className="time-line"
    >
      <span
        className="time-line-progress-bar"
        style={
          Object {
            "width": "33%",
          }
        }
      />
      <input
        onChange={[Function]}
        onInput={[Function]}
        step="0.01"
        type="range"
        value={33}
      />
    </div>
  </div>
</div>
`;

exports[`ProgressBar renders correctly without the progress bar its currentTime 1`] = `
<div
  className="progress"
>
  <div>
    <div
      className="time-line"
    >
      <span
        className="time-line-progress-bar"
        style={
          Object {
            "width": "33%",
          }
        }
      />
      <input
        onChange={[Function]}
        onInput={[Function]}
        step="0.01"
        type="range"
        value={33}
      />
    </div>
  </div>
  <div
    className="counters"
  >
    <span>
      3u 25m 45s
    </span>
  </div>
</div>
`;

exports[`ProgressBar renders correctly without the progress bar its duration 1`] = `
<div
  className="progress"
>
  <div>
    <div
      className="time-line"
    >
      <span
        className="time-line-progress-bar"
        style={
          Object {
            "width": "33%",
          }
        }
      />
      <input
        onChange={[Function]}
        onInput={[Function]}
        step="0.01"
        type="range"
        value={33}
      />
    </div>
  </div>
  <div
    className="counters"
  >
    <span>
      20m 34s
    </span>
  </div>
</div>
`;

exports[`ProgressBar renders correctly without time-line 1`] = `
<div
  className="progress"
>
  <div />
  <div
    className="counters"
  >
    <span>
      20m 34s
    </span>
    <span>
      3u 25m 45s
    </span>
  </div>
</div>
`;
